content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
+
+As shown in Figure 7, the user will pass the initial program $x\_{0}$ into the BLOKE optimizer, which would put the program into the in\_queue of phase pipeline. In phase 1 , we spawn off a predetermined number of sampling processes on the initial program.
+
+Every time the sampling process of phase $n$ found a better optimization $x^{\prime}$, the process will pass $x^{\prime}$ into the phase $n$ output queue to which phase $n+1$ consumes with their own sampling processes.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-06.jpg?height=588&width=805&top_left_y=1191&top_left_x=237)
+
+Figure 7: BLOKE parallelization
+
+## Evaluation
+
+We implemented BLOKE in 3200 lines of Python. We chose to not use the existing STOKE library because $\mathrm{x} 86$ assembly code logic was deeply entangled in its codebase.
+
+Sadly, not many existing Bril benchmarks can be optimized with BLOKE currently since some opcodes are not supported. BLOKE also cannot optimize programs with loops yet, which is an item for future work. So to evaluate
+BLOKE, we made short toy examples shown in Figure 8. clobber is shown in Figure 1.
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-07.jpg?height=272&width=362&top_left_y=304&top_left_x=232)
+
+Figure 8a: distribute
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-07.jpg?height=303&width=344&top_left_y=263&top_left_x=674)
+
+Figure 8b: dead\_code
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-07.jpg?height=273&width=404&top_left_y=270&top_left_x=1110)
+
+Figure 8c: unused\_br
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-07.jpg?height=309&width=338&top_left_y=234&top_left_x=1538)
+
+Figure 8d: dead\_code\_alt
+
+We measured our benchmarks on a Dell PowerEdge R620 with $62.8 \mathrm{GiB}$ of memory and 32 3.4GHz Intel Xeon E5-2650 v2 cores running Ubuntu 22.04.3 LTS. Each BLOKE run in our evaluation was run with 10,000 iterations.
+
+For each benchmark, we measured the (1) performance of the optimizations BLOKE found, (2) how multiple phases affected the optimizations, (3) how fast BLOKE runs with respect to the number of processes BLOKE can have running at one time, and (4) how the number of phases impacted BLOKE runtime.
+
+## Optimization Performance
+
+To measure the (1) performance of the optimizations BLOKE found, we ran the five benchmarks with 5 BLOKE phases and $\beta \in[1,10]$. Figure 9a shows the total number of dynamic executions executed in the original program in blue, and the number of dynamic executions executed in the optimized program in red. For most of the benchmarks we were able to find optimizations. For the ones that BLOKE found optimizations, they turned out to be the most optimal program.
```

Comment:
For the blog post format, I don't think you need to number your figures… just say "this chart shows…" and then include the image right there.